### PR TITLE
[Draft] Edit: general refactor

### PR DESCRIPTION
**Pull request by carlopav**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [ ] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

File: src/Mod/Draft/DraftEdit.py [hunk 1]
```diff
@@ -244,7 +245,7 @@ def keyPressed(self, event_callback):
             if key == 111: # "o"
                 self.finish(closed=True)
             if key == 105: # "i"
-                if Draft.getType(self.obj) == "Circle": self.arcInvert()
+                if Draft.getType(self.obj) == "Circle": self.arcInvert(self.obj)
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: I suggest complying to PEP8, so avoid control structures in the same line. It's better to break them to two lines. More readable this way.
```py
if Draft.getType(self.obj) == "Circle":
    self.arcInvert(self.obj)
```

File: src/Mod/Draft/DraftEdit.py [hunk 2]
```diff
@@ -447,21 +448,26 @@ def getEditNodeIndex(self, point):
     # EDIT TRACKERS functions
     #---------------------------------------------------------------------------
 
-    def setTrackers(self, obj):
+    def setTrackers(self, obj, points=None):
         "set Edit Trackers for editpoints collected from self.obj"
+        if points == None or len(points) == 0:
+            FreeCAD.Console.PrintWarning(translate("draft", "No edit point found for selected object")+"\n")
+            # do not finish if some trackers are still present
+            if self.trackers == {'object':[]}: self.finish()
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]:
```py
if self.trackers == {'object':[]}:
    self.finish()
```

File: src/Mod/Draft/DraftEdit.py [hunk 3]
```diff
@@ -706,78 +712,76 @@ def delPoint(self,event):
...
+        elif Draft.getType(obj) == "Dimension": self.updateDimension(obj, nodeIndex, v)
+        elif Draft.getType(obj) == "Sketch": self.updateSketch(obj, nodeIndex, v)
+        elif Draft.getType(obj) == "Wall": self.updateWall(obj, nodeIndex, v)
+        elif Draft.getType(obj) == "Window": self.updateWindow(obj, nodeIndex, v)
+        elif Draft.getType(obj) == "Space": self.updateSpace(obj, nodeIndex, v)
+        elif Draft.getType(obj) == "Structure": self.updateStructure(obj, nodeIndex, v)
+        elif Draft.getType(obj) == "PanelCut": self.updatePanelCut(obj, nodeIndex, v)
+        elif Draft.getType(obj) == "PanelSheet": self.updatePanelSheet(obj, nodeIndex, v)
+        elif Draft.getType(obj) == "Part::Line" and self.obj.TypeId == "Part::Line": self.updatePartLine(obj, nodeIndex, v)
+        elif Draft.getType(obj) == "Part" and self.obj.TypeId == "Part::Box": self.updatePartBox(obj, nodeIndex, v)
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: Again, even if the old code is in the same line. It's more readable if each branch is in a separate line.

File: src/Mod/Draft/DraftEdit.py [hunk 4]
```diff
@@ -903,25 +909,25 @@ def resetTrackersBezier(self, obj):
...
 
-    def smoothBezPoint(self, point, style='Symmetric'):
+    def smoothBezPoint(self, obj, point, style='Symmetric'):
         "called when changing the continuity of a knot"
         style2cont = {'Sharp':0,'Tangent':1,'Symmetric':2}
         if point is None: return
-        if not (Draft.getType(self.obj) == "BezCurve"):return
-        pts = self.obj.Points
-        deg = self.obj.Degree
+        if not (Draft.getType(obj) == "BezCurve"):return
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]:
```py
if point is None:
    return
if Draft.getType(obj) != "BezCurve":
    return
```

File: src/Mod/Draft/DraftEdit.py [hunk 5]
```diff
@@ -932,10 +938,10 @@ def smoothBezPoint(self, point, style='Symmetric'):
                     return
                 if knot:
                     if style == 'Tangent':
-                        pts[changep] = self.obj.Proxy.modifytangentpole(\
+                        pts[changep] = obj.Proxy.modifytangentpole(\
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: No need to use backslash inside parentheses, as the parentheses break the line correctly.
```py
pts[changep] = obj.Proxy.modifytangentpole(pts[knot], pts[keepp], pts[changep])
```

Or
```py
pts[changep] = obj.Proxy.modifytangentpole(pts[knot],
                                           pts[keepp], pts[changep])
```

File: src/Mod/Draft/DraftEdit.py [hunk 6]
```diff
@@ -932,10 +938,10 @@ def smoothBezPoint(self, point, style='Symmetric'):
...
                                 pts[knot],pts[keepp],pts[changep])
                     elif style == 'Symmetric':
-                        pts[changep] = self.obj.Proxy.modifysymmetricpole(\
+                        pts[changep] = obj.Proxy.modifysymmetricpole(\
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: Same here.

File: src/Mod/Draft/DraftEdit.py [hunk 7]
```diff
@@ -944,83 +950,79 @@ def smoothBezPoint(self, point, style='Symmetric'):
...
+                    pts[1],pts[-1] = obj.Proxy.symmetricpoles(pts[0],pts[1],pts[-1])
                 knot = 0
             else:
                 FreeCAD.Console.PrintWarning(translate("draft", "Endpoint of BezCurve can't be smoothed")+"\n")
                 return
         segment = knot // deg #segment index
-        newcont=self.obj.Continuity[:] #don't edit a property inplace !!!
-        if not self.obj.Closed and (len(self.obj.Continuity) == segment -1 or \
+        newcont=obj.Continuity[:] #don't edit a property inplace !!!
+        if not obj.Closed and (len(obj.Continuity) == segment -1 or \
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: No backslash inside parentheses.
```py
if not obj.Closed and (len(obj.Continuity) == segment-1 or  segment == 0):
    pass # open curve
```
or
```py
if not obj.Closed and (len(obj.Continuity) == segment-1
                       or  segment == 0):
    pass # open curve
```

File: src/Mod/Draft/DraftEdit.py [hunk 8]
```diff
@@ -1136,23 +1143,35 @@ def updateCircle(self,v):
...
+    
+    def getArcEnd(self, obj):#Returns object midpoint
+        if Draft.getType(obj) == "Circle":
+            return self.pointOnCircle(obj, obj.LastAngle)
 
     def getArcMid(self, obj):#Returns object midpoint
         if Draft.getType(obj) == "Circle":
             if obj.LastAngle>obj.FirstAngle:
-                midAngle=math.radians(obj.FirstAngle+(obj.LastAngle-obj.FirstAngle)/2)
+                midAngle=obj.FirstAngle+(obj.LastAngle-obj.FirstAngle)/2
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: Divisions should use a floating point number, to make sure the result is a float. This is important in Python 2. In Python 3 is not needed but it doesn't hurt. Also add spaces between the operators for readability.
```py
midAngle = obj.FirstAngle + (obj.LastAngle - obj.FirstAngle)/2.0
```

File: src/Mod/Draft/DraftEdit.py [hunk 9]
```diff
@@ -1209,62 +1232,97 @@ def updateDimension(self,v):
...
         if obj.GeometryCount == 1:
-            self.editpoints.append(self.pl.multVec(obj.getPoint(0,1)))
-            self.editpoints.append(self.pl.multVec(obj.getPoint(0,2)))
+            editpoints.append(obj.getGlobalPlacement().multVec(obj.getPoint(0,1)))
+            editpoints.append(obj.getGlobalPlacement().multVec(obj.getPoint(0,2)))
+            return editpoints
         else:
             FreeCAD.Console.PrintWarning(translate("draft","Sketch is too complex\
-                 to edit: it is suggested to use sketcher editor")+"\n")
+                 to edit: it is suggested to use sketcher default editor")+"\n")
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: Continuation inside the parentheses can be done with strings as well, without the backslash.

```py
FreeCAD.Console.PrintWarning(translate("draft",
                                       "Sketch is too complex"
                                       "to edit: it is suggested to use sketcher default editor")+"\n")
```

**Comment by carlopav** [comment]
Thx for reviewing, I'll fix them all!

**Comment by carlopav** [comment]
Done!

File: src/Mod/Draft/DraftEdit.py [hunk 10]
```diff
@@ -288,8 +309,9 @@ def mouseMoved(self, event_callback):
             pos = event.getPosition()
             node = self.getEditNode(pos)
             ep = self.getEditNodeIndex(node)
-            if ep != None: 
-                if self.overNode != None: self.overNode.setColor(COLORS["default"])
+            if ep != None:
+                if self.overNode != None:
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: Testing for `None` uses the `is` operator.
```py
if ep is not None:
    if self.overNode is not None:
```

Comment (carlopav) [reply src/Mod/Draft/DraftEdit.py]: i'll include it in next commit. Thanks!

File: src/Mod/Draft/DraftEdit.py [hunk 11]
```diff
@@ -302,14 +324,17 @@ def startEditing(self, event):
...
         doc = FreeCAD.getDocument(str(node.documentName.getValue()))
         self.obj = doc.getObject(str(node.objectName.getValue()))
-        if self.obj is None: return
+        if self.obj is None:
+            return
         self.setPlacement(self.obj)
 
-        FreeCAD.Console.PrintMessage(str(self.obj.Name)+str(": editing node: n° ")+str(ep)+"\n")
+        FreeCAD.Console.PrintMessage(str(self.obj.Name) + str(
+            ": editing node: n° ") + str(ep) + "\n")
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: You can break here at the operator, as it's already inside a pair of parentheses.
```py
FreeCAD.Console.PrintMessage(str(self.obj.Name)
                             + str(": editing node: n° ")
                             + str(ep) + "\n")
```

File: src/Mod/Draft/DraftEdit.py [hunk 12]
```diff
@@ -57,9 +58,13 @@
...
-
-    "The Draft_Edit FreeCAD command definition"
-
+    """
+    The Draft_Edit FreeCAD command definition.
+    A tool to graphically edit FreeCAD objects.
+    Current implementation use many parts of pivy graphics code by user "looo".
+    The tool collect editpoints from objects and display Trackers on them to allow
+    editing their Shape and their parameters.
+    """
```
Comment (vocx-fc) [inline src/Mod/Draft/DraftEdit.py]: I see that you added a lot of documentation strings. That's good.

For the class, you can add a short explanation of the attributes, that is, all those variables like `self.running`, `self.trackers`, etc., without the word `self.`

```py
Class Edit():
    """
    Attributes
    ------
    running : bool
        Whether the tool is running or not.
    trackers : dictionary
        The dictionary that holds the current object and its tracker.
    ...
    """
```

Comment (carlopav) [reply src/Mod/Draft/DraftEdit.py]: Yes! i'll do it

Comment (vocx-fc) [reply src/Mod/Draft/DraftEdit.py]: Verbosity is not a problem. As a matter of fact we want the docstrings to be verbose. Then we don't have to write separate documentation, we can just take a look at the source. Tools like Doxygen and Sphinx parse these documentation strings and produce good documentation that can be used by both users and programmers.

Take a look at WorkingPlane.py to get an idea how I wrote the documentation strings there.

**Pull request merged**